src/sentry/integrations/gitlab/webhooks.py

```diff
@@ -193,75 +193,101 @@ def dispatch(self, request: Request, *args, **kwargs) -> Response:
+                },
+            }
+        except Integration.DoesNotExist as e:
+            logger.info("gitlab.webhook.invalid-organization", extra=extra)
+            extra["reason"] = "There is not integration that matches your organization."
+            capture_exception(e)
+            return HttpResponse(status=400, reason=extra["reason"])
 
-        if not constant_time_compare(secret, integration.metadata["webhook_secret"]):
-            logger.info(
-                "gitlab.webhook.invalid-token-secret", extra={"integration_id": integration.id}
-            )
-            # XXX: Look into using set_context
-            capture_message("Gitlab's webhook secret does not match.")
-            return HttpResponse(
-                status=400,
-                reason="Gitlab's webhook secret does not match. Refresh token (or re-install the integration) by following this https://docs.sentry.io/product/integrations/integration-platform/public-integration/#refreshing-tokens.",
-            )
+        try:
+            if not constant_time_compare(secret, integration.metadata["webhook_secret"]):
+                # This forces a stack trace to be produced
+                raise Exception("The webhook secrets do not match.")
+        except Exception as e:
+            logger.info("gitlab.webhook.invalid-token-secret", extra=extra)
+            extra[
+                "reason"
+            ] = "Gitlab's webhook secret does not match. Refresh token (or re-install the integration) by following this https://docs.sentry.io/product/integrations/integration-platform/public-integration/#refreshing-tokens."
+            capture_exception(e)
+            return HttpResponse(status=400, reason=extra["reason"])
 
         try:
             event = json.loads(request.body.decode("utf-8"))
         except json.JSONDecodeError as e:
-            logger.info(
-                "gitlab.webhook.invalid-json", extra={"external_id": integration.external_id}
-            )
+            logger.info("gitlab.webhook.invalid-json", extra=extra)
+            extra["reason"] = "Data received is not JSON."
             capture_exception(e)
-            return HttpResponse(status=400, reason="Data received is not JSON.")
+            return HttpResponse(status=400, reason=extra["reason"])
 
         try:
             handler = self._handlers[request.META["HTTP_X_GITLAB_EVENT"]]
         except KeyError as e:
-            logger.info(
-                "gitlab.webhook.missing-event", extra={"event": request.META["HTTP_X_GITLAB_EVENT"]}
-            )
+            logger.info("gitlab.webhook.wrong-event-type", extra=extra)
+            supported_events = ", ".join(sorted(self._handlers.keys()))
+            logger.info(f"We only support these kinds of events: {supported_events}")
+            extra[
+                "reason"
+            ] = "The customer has edited the webhook in Gitlab to include other types of events."
```

Comment:
An example of this is [this issue](https://sentry.io/organizations/sentry/issues/3610339170/?query=is%3Aunresolved+gitlab&statsPeriod=24h).
<img width="545" alt="image" src="https://user-images.githubusercontent.com/44410/191786867-f5bb6915-2623-42c7-adc7-e92a159a54b2.png">